src/utils/queryOps.js

```diff
@@ -1,225 +1,416 @@
+          output[skey] = triple;
+        }
+
+        triple[0].push(arg[0]);
+      }
+    } else {
+      // single-column groupby is faster
+      let g = 0;
+
+      for (const [row, ...arg] of this.argslist) {
+        const key = func(row, ...arg, g);
+        let triple = output[key];
+
+        if (!triple) {
+          triple = [[], key, g];
+          g += 1;
+          output[key] = triple;
+        }
+
+        triple[0].push(row);
+      }
+    }
+
+    return new Wrapper(Object.values(output));
+  }
+
+  // ///////////////////////////////////////////////////////////////////////////
+  // TERMINAL METHODS
+  // ///////////////////////////////////////////////////////////////////////////
+
+  toArray() {
+    return Array.from(this);
+  }
+
+  fromPairs() {
+    // return an object from (value, key) pairs
+    const output = {};
+
+    for (const [v, k] of this.argslist) {
+      output[k] = v;
+    }
+
+    return output;
+  }
+
+  first() {
+    // return first element
+    return first(this);
+  }
+
+  findIndex(func) {
+    // return index of first element where func returns true
+    // return null if not found
+    let i = 0;
+
+    for (const args of this.argslist) {
+      if (func(...args)) return i;
+      i += 1;
+    }
+
+    return null;
+  }
+
+  last() {
+    // return last element
+    return last(this);
+  }
+
+  index(column) {
+    // Return an object indexed on `column`
+    // We assume the key is unique
+    const output = {};
+
+    for (const [row] of this.argslist) {
+      const key = row[column];
+
+      if (!(key in output)) {
+        output[key] = row;
+      } else {
+        throw new Error('expecting index to be unique');
+      }
+    }
+
+    return output;
+  }
+}
+
+function frum(list) {
+  if (list instanceof Wrapper) {
+    return list;
+  }
+
+  if (Array.isArray(list)) {
+    return new Wrapper(list.map(v => [v]));
+  }
+
+  return new Wrapper(Array.from(list));
+}
+
+internalFrum = frum;
+
+function toPairs(obj) {
+  // convert Object (or Map) into [value, key] pairs
+  // notice the **value is first**
+  if (obj instanceof Map) {
+    return new Wrapper(Array.from(obj.entries()).map(([k, v]) => [v, k]));
+  }
+
+  return new Wrapper(Object.entries(obj).map(([k, v]) => [v, k]));
+}
+
+internalToPairs = toPairs;
+
+function length(list) {
+  // return length of this list
+  if (list instanceof Wrapper) return list.length;
+
+  return list.length;
+}
+
+function extendWrapper(methods) {
+  // Add a chainable method to Wrapper
+  internalToPairs(methods).forEach((method, name) => {
+    // USE function(){} DECLARATION TO BIND this AT CALL TIME
+    Wrapper.prototype[name] = function anonymous(...args) {
+      return internalFrum(method(this.toArray(), ...args));
+    };
+  });
+}
+
+// Add Lodash functions
+extendWrapper({
+  chunk,
+  unzip,
+  zip: unzip,
+  limit: lodashTake,
+  sort: sortBy,
+  sortBy,
+
+  // SELECT a.*, b.* FROM listA a LEFT JOIN listB b on b[propB]=a[propA]
```

Comment:
Since this is an open-source frontend project, it'd be nice to not assume SQL knowledge have either an explanation or a commented out representation of the data structure.